pgjdbc/src/main/java/org/postgresql/xa/PGXAConnection.java

```diff
@@ -395,13 +407,20 @@ public void rollback(Xid xid) throws XAException {
           stmt.close();
         }
       }
+      committedOrRolledBack = true;
     } catch (SQLException ex) {
+      int errorCode = XAException.XAER_RMERR;
       if (PSQLState.UNDEFINED_OBJECT.getState().equals(ex.getSQLState())) {
-        throw new PGXAException(GT.tr("Error rolling back prepared transaction"), ex,
-            XAException.XAER_NOTA);
+        if (committedOrRolledBack || !xid.equals(preparedXid)) {
+          // working with unknown xid or already committed/rolled-back
+          errorCode = XAException.XAER_NOTA;
+        }
       }
-      throw new PGXAException(GT.tr("Error rolling back prepared transaction"), ex,
-          XAException.XAER_RMERR);
+      // connection failure, reconnection could be expected
+      if (PSQLState.isConnectionError(ex.getSQLState())) {
+        errorCode = XAException.XAER_RMFAIL;
+      }
+      throw new PGXAException(GT.tr("Error rolling back prepared transaction"), ex, errorCode);
```

Comment:
@ochaloup , I might be a bit late to the party, however what do you think of printing things like `xid`, `preparedXid`, `committedOrRolledBack` in the error message?

I'm sure, that would be crucial information if somebody receives `""Error rolling back prepared transaction"`.

Reply:
@vlsi : I agree. Just I was not sure if I can change the format as those information was not part of the messages previously. I'm going to add the proposed info to the messages, push it to the PR and please check that then again.